dali/operators/decoder/audio/generic_decoder.cc

```diff
@@ -0,0 +1,230 @@
+// Copyright (c) 2019, NVIDIA CORPORATION. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "dali/operators/decoder/audio/generic_decoder.h"
+
+namespace dali {
+
+namespace {
+
+template<typename SampleType>
+size_t read_samples(SNDFILE *snd_file, span<SampleType> output) {
+  DALI_FAIL("Can't find function for given type. You may want to write your own specialization.")
+}
+
+
+template<>
+size_t read_samples<short>(SNDFILE *snd_file, span<short> output) {  // NOLINT
+  return sf_read_short(snd_file, output.data(), output.size());
+}
+
+
+template<>
+size_t read_samples<int>(SNDFILE *snd_file, span<int> output) {
+  return sf_read_int(snd_file, output.data(), output.size());
+}
+
+
+template<>
+size_t read_samples<float>(SNDFILE *snd_file, span<float> output) {
+  return sf_read_float(snd_file, output.data(), output.size());
+}
+
+
+template<>
+size_t read_samples<double>(SNDFILE *snd_file, span<double> output) {
+  return sf_read_double(snd_file, output.data(), output.size());
+}
+
+
+class MemoryStream {
+ public:
+  MemoryStream() : length_(0), curr_(0), input_(nullptr) {}
+
+
+  MemoryStream(const void *input, int length /* bytes */) :
+          length_(length), curr_(0), input_(static_cast<const char *>(input)) {}
+
+
+  size_t length() {
+    return length_;
+  }
+
+
+  size_t seek(sf_count_t offset, int whence) {
+    switch (whence) {
+      case SEEK_SET:
+        curr_ = offset;
+        break;
+      case SEEK_CUR:
+        curr_ += offset;
+        break;
+      case SEEK_END:
+        curr_ = length_ + offset;
+        break;
+      default:
+        DALI_FAIL("Incorrect `whence` argument")
+    }
+    return curr_;
+  }
+
+
+  size_t read(void *dst, sf_count_t num) {
+    num = std::min<sf_count_t>(num, length_ - curr_);
+    memcpy(dst, input_ + curr_, num);
+    curr_ += num;
+    return num;
+  }
+
+
+  size_t tell() {
+    return curr_;
+  }
+
+
+ private:
+  size_t length_, curr_;
+  const char *input_;
+};
+
+/*
+ * Functions: `GetFileLen`, `Seek`, `Read`, `Tell` are the callbacks required
+ * by `sf_open_fd` interface. Descriptions of these functions are copy-pasted
+ * from `libsnd` docs.
+ */
+
+/**
+ * The virtual file context must return the length of the virtual file in bytes.
+ */
+sf_count_t GetFileLen(void *This) {
```

Comment:
I don't thing we should use capitalized names for variable names. Maybe `p_this` or `self`

Reply:
Done